from PyQt5.uic import loadUi
from re import compile, IGNORECASE

from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import QWidget

from views.misc.messagebox import MessageBox


class GeneratorView(QWidget):
    def __init__(self, parent):
        super(GeneratorView, self).__init__(parent)
        loadUi('./stuff/ui/exploits/generator.ui', self)
        self.exploit = None
        self.category = "generator"

        # Connect buttons
        self.pushButtonGenerate.clicked.connect(self.pushButtonGenerateClicked)
        self.labelAuthor.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelDate.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelCWE.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelTarget.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelDescription.setTextInteractionFlags(Qt.TextSelectableByMouse)
        self.labelDescription.setWordWrap(True)

        self.labelReferences.setTextFormat(Qt.RichText)
        self.labelReferences.setTextInteractionFlags(Qt.TextBrowserInteraction)
        self.labelReferences.setOpenExternalLinks(True)

        # Set verticalLayout. It's resize the widget
        self.setLayout(self.verticalLayout)

    def check(self):
        mac = self.lineEditMAC.text()
        # Credits http://stackoverflow.com/questions/18782926/regex-to-match-mac-address-and-also-extract-its-values
        regex = compile(r'^([0-9A-F]{1,2})\:([0-9A-F]{1,2})\:([0-9A-F]{1,2})\:([0-9A-F]{1,2})\:([0-9A-F]{1,2})\:([0-9A-F]{1,2})$', IGNORECASE)
        result = regex.match(mac)

        if not result:
            MessageBox.critical("Error", "You have to set a MAC. Example: '00:11:22:33:44:55'")
            return False
        else:
            return True

    def getCategory(self):
        return self.category

    def pushButtonGenerateClicked(self):
        if self.check():
            mac = self.lineEditMAC.text()
            try:
                data = self.exploit.run(mac)
            except Exception as e:
                data = e.message

            self.labelOutput.setText(str(data))

    def setExploit(self, exploit):
        self.exploit = exploit
        self.labelNameExploit.setText(self.exploit.getName())
        self.labelAuthor.setText(self.exploit.getAuthors())
        self.labelDate.setText(self.exploit.getDate())
        self.labelTarget.setText(self.exploit.getTarget())
        self.labelCWE.setText(self.exploit.getCWE())
        self.labelReferences.setText(self.exploit.getReferences())
        self.labelDescription.setText(self.exploit.getDescription())
        self.lineEditMAC.setText(self.exploit.getMAC())
        self.labelOutput.setText("")
